mitodl/mitxonline
# Adds additional demographic fields to the system; adds popup to collect more data when visiting a course

#### Pre-Flight checklist


#### What are the relevant tickets?

Closes #1424 

#### What's this PR do?

Adds a number of demographic data fields to the user profile. These expose the fields added to the database in [#1443](https://github.com/mitodl/mitxonline/pull/1443). 

Adds a popup requesting more data from the learner when they attempt to navigate to a course. This popup only occurs once and includes a form that has the new fields from #1443 and the Gender and Year of Birth fields. 

Adds a utility function to the front end code that should make working with feature flags easier in the future. 

#### How should this be manually tested?

Automated tests should pass.

All new UI elements are hidden behind a feature flag - set `FEATURE_ENABLE_ADDL_PROFILE_FIELDS` to True in your .env file to test. Alternatively, append `?enable_addl_profile_fields` to the URL on the dashboard and profile pages to see the changes. 

Testing profile form:
1. Log in as a user. 
2. Navigate to the profile page.
3. Observe the fields there. You should see now a set of checkboxes listed - one each for Educator, Professional, Student, and Other. None of them should be required.
4. Checking them and saving should work.
5. Checking Professional should expand the form to include: 
   * Company
   * Job Title
   * Industry
   * Job Function
   * Company Size
   * Years of Work Experience
   * Leadership Level
   * Highest Level of Education
6. Filling out these fields should work.

Testing popup:
You will need a user that's enrolled in a course (enrollment mode doesn't matter here; they just need to be enrolled at all). 
1. In a separate tab, open the Django Admin and find the user you're testing with. Open the User Profile section and observe the status of the `Addl field flag` field. 
2. _If the flag is set_: navigate to the user's dashboard and click the title of a course to open the course homepage. This should work as expected.
3. _If the flag is **un**set_: navigate to the user's dashboard and click the title of a course to open the course homepage.
   1. Reloading the user's record in the Django Admin should now show the flag is set. 
   2. A "Provide more info" dialog should appear, requesting more information. The fields should be Gender, Year of Birth, and "Are you a". The Professional checkbox should show and hide the additional company information as it does on the Profile page.
   3. Clicking Save should save the information into the profile and continue on to open the course homepage. Note that since the dialog will do this programmatically, your browser may block the popup. 

#### Screenshots (if appropriate)
Edit Profile (Professional not checked):
![image](https://user-images.githubusercontent.com/945611/221236826-b715d4e5-7053-45c9-a49f-4a893763210c.png)

Edit Profile (Professional checked):
![image](https://user-images.githubusercontent.com/945611/221236987-e0be5a55-3468-4fe8-94a6-47ca00bb9f81.png)

More Info dialog (Professional not checked):
![image](https://user-images.githubusercontent.com/945611/221237180-28711894-e09b-4839-a521-e188f52d8c02.png)

More Info dialog (Professional checked):
![image](https://user-images.githubusercontent.com/945611/221237256-88ee5764-f46b-4a63-936d-d69227e1f227.png)

More Info dialog (Professional checked; on mobile):
![image](https://user-images.githubusercontent.com/945611/221237459-880d5af1-cb1a-4478-8805-f85962b7b244.png)

Edit Profile (Professional checked; on mobile):
![image](https://user-images.githubusercontent.com/945611/221237673-4ce4f401-acb2-430a-af92-1a7bf505fb0e.png)
